jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java

```diff
@@ -332,29 +372,105 @@ public String toString()
         return getClass().getSimpleName() + " size=" + sharedList.size() + " closed=" + closed + " entries=" + sharedList;
     }
 
+    /**
+     * A Reservation of a slot in the Pool
+     */
+    public class Reservation
+    {
+        private final Entry entry = new Entry();
+
+        /**
+         * @return The reserved {@link Entry}, which will be closed until enabled.
+         */
+        public Pool<T>.Entry getEntry()
+        {
+            return entry;
+        }
+
+        /** Enable the reserved {@link Entry}.
+         * Once enabled, the entry is immediately available to be acquired, potentially by
+         * another thread.  If the caller wishes to acquire the associated entry, they should
+         * use {@link #acquire(Object)} to atomically enable and acquire.
+         * @param pooled The pooled item for the entry
+         */
+        public void enable(T pooled)
+        {
+            enable(pooled, false);
+        }
+
+        /** Enable and acquire the reserved {@link Entry}.
+         * The associated entry is atomically enabled and acquired, so that no other thread can acquire it and
+         * the {@link #getEntry()} value may be used by the caller.
+         * @param pooled The pooled item for the entry
+         */
+        public Pool<T>.Entry acquire(T pooled)
+        {
+            enable(pooled, true);
+            return entry;
+        }
+
+        private void enable(T pooled, boolean acquire)
+        {
+            Objects.requireNonNull(pooled);
+            if (entry.state.getHi() != Integer.MIN_VALUE)
+                throw new IllegalStateException("Open entries cannot be enabled : " + this);
+            entry.pooled = pooled;
+            int usage = acquire ? 1 : 0;
+            if (!entry.state.compareAndSet(Integer.MIN_VALUE, usage, 0, usage))
+            {
+                entry.pooled = null;
+                throw new IllegalStateException("Entry cannot be enabled : " + this);
```

Comment:
A `Reservation` may be concurrently removed while it's being enabled.
This code should be converted to the classic `while (true)` loop and retry in case of failures.
Furthermore, it needs to account for the closed state (`hi==-1`) and not enable if it's already closed.
Which means `enable()` should return a `boolean` that tells whether the enabling could be done.

Reply:
Unfortunately this code is just not threadsafe and I can't see how we can make it so without adding a lock.
The problem is that if we allow multiple threads to call enable on the reservation, both may pass the MIN_VALUE check and then both will try to set the `entry.pooled` value.  Only one will win and it may not be the one that wins the CaS race.

So currently enable will ISE atomically if anybody else messes with state (of the reserved entry or by closing the pool).  The state of the pooled field is not so well defined for multiple threads.... but I think any app that is racing itself to call enable is just wrong anyway.

I think ISE is OK for -1 state.... but are you saying that if it is -1, we should noop?       If so, I can do that after a failed CaS, but not in a loop, which will never work because of the pooled field problem.

I've gone back on forth on this method a lot... I don't like it, but I like it more than a big lock.

Reply:
> The problem is that if we allow multiple threads to call enable on the reservation

But we never supported that and I think it's ok - `enable()` has a natural race with `remove()` but not with itself.

It is possible to write the classic `while (true)` loop, not care about races with itself, but do care about races with `remove()` that are natural and should not throw.

Reply:
The class `while (true)` loop doesn't work if we throw ISE if the state is not what we expected.